### PR TITLE
Use custom classes for user_id and group_id

### LINKED ISSUE #7: Convert user/group ids to strings
As input (e.g. `file_chown()`) and output (e.g. `file_info()`)

### PR DESCRIPTION
**Pull request by jimhester**
To display the ids as characters from file_info() and support supplying
them as names to file_chown()

Fixes #7 

On windows both user_id and group_id are always -1, so the C++ functions are just dummies and are ifndef'd out.

@hadley does this seem like the right approach? I wanted to preserved the ability to reference the ids as either a name or the numeric id. I couldn't use named vectors directly because the names are removed when you put them in a data.frame, and I could not store the names as an attribute because of the dplyr subsetting issue. So I settled on this mapping approach, where the full map is stored in the object, and the values are translated as needed by the methods.

### CODE REVIEW

**Comment by jimhester** [comment]
This still needs some tests, which are not entirely straightforward because we do not know what users or groups will be available on the systems the tests are run.

File: R/file.R [hunk 1]
```diff
@@ -92,9 +95,10 @@ file_copy <- function(path, new_path, overwrite = FALSE) {
 
 #' Change ownership or group of a file
 #' @template fs
-#' @param user_id The user id of the new owner, the R process must be privlaged to change
-#'   this.
-#' @param group_id The group id of the new owner
+#' @param user_id The user id of the new owner, specified as a numeric ID or
+#'   name. The R process must be privileged to change this.
```
Comment (hadley) [inline R/file.R]: BTW I don't know what "R process must be privileged" means

File: R/file.R [hunk 2]
```diff
@@ -92,9 +95,10 @@ file_copy <- function(path, new_path, overwrite = FALSE) {
 
 #' Change ownership or group of a file
 #' @template fs
-#' @param user_id The user id of the new owner, the R process must be privlaged to change
-#'   this.
-#' @param group_id The group id of the new owner
+#' @param user_id The user id of the new owner, specified as a numeric ID or
```
Comment (hadley) [inline R/file.R]: Are numeric IDs only valid on windows?

Comment (jimhester) [reply R/file.R]: no ids are valid on windows, both uid and gid just return -1 on windows.

File: R/file.R (hunk 1 above)
Comment (jimhester) [reply R/file.R]: It essentially means you have to run the R process as a super user (either as root or su) to change it (https://en.wikipedia.org/wiki/Privilege_(computing)#Unix)

Comment (hadley) [reply R/file.R]: So this means that these functions are of relatively little use then (generally)?  I think that means we should just do the minimum here and wait for user requests before going deeper.

**Comment by jimhester** [comment]
An alternative would to provide both the username and the id in separate columns in the data.frame. But I think I have been persuaded that we can just return them as a regular character vector and provide a lookup function (which I already had to write for this anyway).

**Comment by jimhester** [comment]
Ok I have switched to always returning character names for the user and group. On windows it just returns NA for both as this information does not exist there.

File: R/id.R [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#' Lookup Users and Groups on a system
+#'
+#' These functions use the GETPWENT(3) and GETGRENT(3) system calls to query
+#' users and groups respectively. They return their results in a `data.frame`.
```
Comment (hadley) [inline R/id.R]: Use `@return` here?

File: R/id.R [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+#'
+#' On windows both functions return an empty data.frame.
```
Comment (hadley) [inline R/id.R]: Because...

**Comment by jimhester** [comment]
I merged this in https://github.com/r-lib/fs/commit/56ee6ed1c7c9dd692c3b0238744337276b37321b, but GH did not pick it up. Incorporated your documentation suggestions in https://github.com/r-lib/fs/commit/59760fa86b89d3b0a326023843fa0e569d478b8a, thanks for reviewing!

**Pull request closed without merging**